src/node.cc

```diff
@@ -1296,86 +1296,160 @@ void AddPromiseHook(v8::Isolate* isolate, promise_hook_func fn, void* arg) {
   env->AddPromiseHook(fn, arg);
 }
 
-
-MaybeLocal<Value> MakeCallback(Environment* env,
-                               Local<Value> recv,
-                               const Local<Function> callback,
-                               int argc,
-                               Local<Value> argv[],
-                               async_context asyncContext) {
+class InternalCallbackScope {
+ public:
+  InternalCallbackScope(Environment* env,
+                        v8::Local<v8::Value> object,
+                        const async_context& asyncContext);
+  ~InternalCallbackScope();
+  void Close();
+
+  inline bool Failed() const { return failed_; }
+  inline void MarkAsFailed() { failed_ = true; }
+  inline v8::Local<v8::Value> GetFailureReturnValue() const {
+    return failure_return_value_;
+  }
+  inline bool IsInnerMakeCallback() const {
+    return callback_scope_.in_makecallback();
+  }
+
+ private:
+  Environment* env_;
+  async_context async_context_;
+  v8::Local<v8::Object> object_;
+  Environment::AsyncCallbackScope callback_scope_;
+  bool failed_ = false;
+  bool pushed_ids_ = false;
+  bool closed_ = false;
+
+  // This is only for backwards compatibility, semantically it's always empty.
+  v8::Local<v8::Value> failure_return_value_;
+};
+
+CallbackScope::CallbackScope(Isolate* isolate,
+                             Local<Value> object,
+                             async_context asyncContext)
+  : private_(new InternalCallbackScope(Environment::GetCurrent(isolate),
```

Comment:
The intention here is that userland code won’t have access to the fields of `InternalCallbackScope`, or become dependent on the class layout ABI-wise.

Semantically, `InternalCallbackScope` and `CallbackScope` are identical, and their lifetimes are tied together.